File: backend/src/repositories/user.repository.js
const crypto = require('node:crypto');
const { query } = require('../config/database');

async function findUserByEmail(email) {
  const result = await query(
    'SELECT * FROM users WHERE LOWER(email) = LOWER($1) LIMIT 1',
    [email],
  );

  return result.rows[0] || null;
}

async function createUser({ name, email, passwordHash, role = 'PICKER' }) {
  const id = crypto.randomUUID();
  const result = await query(
    `INSERT INTO users (id, name, email, password_hash, role, is_active, created_at, updated_at)
     VALUES ($1, $2, $3, $4, $5, TRUE, NOW(), NOW())
     RETURNING id, name, email, role`,
    [id, name, email, passwordHash, role],
  );

  return result.rows[0];
}

async function updateLastLogin(userId) {
  await query(
    'UPDATE users SET last_login_at = NOW(), updated_at = NOW() WHERE id = $1',
    [userId],
  );
}

async function findUserById(id) {
  const result = await query(
    'SELECT id, name, email, role, is_active FROM users WHERE id = $1 LIMIT 1',
    [id],
  );

  return result.rows[0] || null;
}

module.exports = {
  findUserByEmail,
  createUser,
  updateLastLogin,
  findUserById,
};

File: backend/repositories/user_repository.py
from __future__ import annotations  # for `User | None` below, on Python 3.9

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from models.user import User


class UserRepository:
    def __init__(self, db: Session) -> None:
        self.db = db

    def find_by_email(self, email: str) -> User | None:
        stmt = select(User).where(func.lower(User.email) == email.lower())
        return self.db.execute(stmt).scalar_one_or_none()

    def find_by_id(self, user_id: str) -> User | None:
        return self.db.get(User, user_id)

    def create(self, *, name: str, email: str, password_hash: str, role: str = "PICKER") -> User:
        user = User(name=name, email=email, password_hash=password_hash, role=role)
        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)
        return user

    def update_last_login(self, user: User) -> None:
        user.last_login_at = func.now()
        self.db.add(user)
        self.db.commit()
